### PR TITLE
Revised module suitable for new NESTExtensionModule interface

### PR DESCRIPTION
**Pull request by heplesser**
This is the first example (in draft state) of the new `NESTExtensionInterface` and `ModuleManager` approach to handling modules.

### CODE REVIEW

**Comment by heplesser** [comment]
Even while the example module still needs a lot of polishing, this one should be reviewed and merged soon so we have one that works with the new NEST mechanism. Then we also need to think about versioning these modules, we need releases that fit the NEST releases.

**Review by terhorstd — CHANGES_REQUESTED** [review]
Good suggestion to switch to `mynest` namespace.
Some small points

File: src/mymodule.cpp [hunk 1]
```diff
@@ -34,95 +29,44 @@
...
-  */
-  register_drop_odd_spike_connection( "drop_odd_synapse" );
-  
+   */
+  mynest::register_drop_odd_spike_connection( "drop_odd_synapse" );
+
+  // Next three not registered for now, since they do not unload on ResetKernel
   // Register connection rule.
-  nest::kernel().connection_manager.register_conn_builder< StepPatternBuilder >( "step_pattern" );
+  //nest::kernel().connection_manager.register_conn_builder< mynest::StepPatternBuilder >( "step_pattern" );
```
Comment (pnbabu) [inline src/mymodule.cpp:62]: @heplesser Could you remove these comments as they can be unloaded now?

Comment (heplesser) [reply src/mymodule.cpp]: Done.

**Review by pnbabu — CHANGES_REQUESTED** [review]
I have one minor comment. 

Also, what did we decide on the TODOs in the CMakeLists.txt file? Shall we remove them now and create an issue to address it later?

**Comment by pnbabu** [comment]
@heplesser ping!

File: src/mymodule.cpp [hunk 2]
```diff
@@ -34,95 +29,44 @@
...
 
   /* Register a synapse type.
-     Give synapse type as template argument and the name as second argument.
-  */
-  register_drop_odd_spike_connection( "drop_odd_synapse" );
-  
+   */
+  mynest::register_drop_odd_spike_connection( "drop_odd_synapse" );
+
+  // Next three not registered for now, since they do not unload on ResetKernel
```
[the remark this answers is not in the document]
Comment (heplesser) [reply src/mymodule.cpp]: @pnbabu Done.

File: CMakeLists.txt [hunk 3]
```diff
@@ -232,10 +201,8 @@ endforeach()
 
 add_subdirectory( src )
 
-# Install header file
-install( FILES ${MODULE_HEADER} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${MODULE_NAME} )
-
-# Install help --- based on doc/CMakeLists.txt
+# Install help --- based on doc/CMakeLists.txt -- TODO: Help
+# generation needs review
```
[the remark this answers is not in the document]
Comment (pnbabu) [reply CMakeLists.txt:205]: @heplesser what did we decide on these TODOs? Shall we remove them now and fix in a later PR?

File: CMakeLists.txt [hunk 4]
```diff
@@ -181,7 +150,7 @@ set( CMAKE_INSTALL_DATADIR "${NEST_INSTALL_DATADIR}" CACHE STRING "read-only arc
 
 include( GNUInstallDirs )
 
-# CPack stuff. Required for target `dist`.
+# CPack stuff. Required for target `dist`. -- TODO: Do we really need this?
```
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:153]: Removed, see #27 for follow-up.

File: CMakeLists.txt [hunk 5]
```diff
@@ -212,15 +181,15 @@ set( CPACK_SOURCE_PACKAGE_FILE_NAME ${MODULE_NAME} )
 set( CPACK_PACKAGE_INSTALL_DIRECTORY "${MODULE_NAME} ${MODULE_VERSION}" )
 include( CPack )
 
-# add make dist target
+# add make dist target -- TODO: Do we want this?
```
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:184]: TODO removed, see #28 for follow-up.

File: CMakeLists.txt [hunk 6]
```diff
@@ -212,15 +181,15 @@ set( CPACK_SOURCE_PACKAGE_FILE_NAME ${MODULE_NAME} )
...
 add_custom_target( dist
     COMMAND ${CMAKE_MAKE_PROGRAM} package_source
     # not sure about this... seems, that it will be removed before dist...
     # DEPENDS doc
     COMMENT "Creating a source distribution from ${MODULE_NAME}..."
     )
 
-# check the user link libraries whether installed or not
+# check the user link libraries whether installed or not -- TODO: need this?
```
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:192]: TODO removed, see #29 for follow-up.

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:205]: TODO removed, see #26 for follow-up.

**Comment by heplesser** [comment]
@terhorstd @pnbabu I have now addressed your comments, I hope to your content.

File: CMakeLists.txt [hunk 7]
```diff
@@ -232,10 +201,8 @@ endforeach()
 
 add_subdirectory( src )
 
-# Install header file
-install( FILES ${MODULE_HEADER} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${MODULE_NAME} )
-
 # Install help --- based on doc/CMakeLists.txt
+# generation needs review
```
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:205]: I removed the comment, covered by #26

**Review by pnbabu — COMMENTED** [review]
It looks good to me overall. I am fine with merging this after fixing the failing build on macOS.

**Comment by heplesser** [comment]
> It looks good to me overall. I am fine with merging this after fixing the failing build on macOS.

I have now hiked the Python version to 3.12, so building should work again on macOS.

**Comment by heplesser** [comment]
But now the Linux build failed with what looks like a problem not related to our code. Strangely, I do not find a re-run button for actions in this repo. Let's see how it goes with the macOS build.

**Comment by pnbabu** [comment]
> But now the Linux build failed with what looks like a problem not related to our code. Strangely, I do not find a re-run button for actions in this repo. Let's see how it goes with the macOS build.

Perhaps it's the access issue? I could see the button and the job is running again now.

**Review by pnbabu — APPROVED** [review]
The CI is green!

**Review by terhorstd — APPROVED** [review]
At some point we will have a discussion about the Python versions again, but that's a different story and will be told in another PR.
Thanks for all the effort!

**Pull request merged**